### DISCUSSION TITLE
Issue with setting active_power in Berry

### DISCUSSION THREAD

**Discussion by goodkiller** [arendst/Tasmota#21913] [General]
I have a few modbus devices which gives out only Voltage and Active Power per phase, no current available. Tasmota could calculate current by Voltage and Power as it calculates Power from Voltage and Current. 
I am using dummy energy driver.

Here is one test example:
```
energy.current_available = 1
energy.frequency_common = 1
energy.active_power
0
energy.power_factor
nan
energy.power_factor = 0
energy.power_factor
0
energy.power_factor = 1
energy.power_factor
1
energy.voltage = 240
energy.voltage
240
energy.active_power = 1000
energy.active_power
0
energy.current
0
energy.current = 10
energy.current
10
energy.active_power
0
energy.active_power = 5555
energy.active_power
2400
energy.update_total(0)
energy.active_power
2400
``` 

And code which is calculating Power from Current:
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_03_esp32_energy.ino
```c
if (!Energy->type_dc) {
    if (Energy->current_available && Energy->voltage_available) {
      for (uint32_t i = 0; i < Energy->phase_count; i++) {
        apparent_power[i] = Energy->apparent_power[i];
        if (isnan(apparent_power[i])) {
          apparent_power[i] = Energy->voltage[i] * Energy->current[i];
        }
        else if (0 == Energy->current[i]) {
          apparent_power[i] = 0;
        }
        if (apparent_power[i] < Energy->active_power[i]) {  // Should be impossible
          Energy->active_power[i] = apparent_power[i];
        }
```

**Comment by s-hadinger** [comment]
I'm trying to reproduce, but I'm not sure what you mean with dummy driver. Is it the one enabled with `OPTION A2`?

In such case, values are set through the driver, not by writing directly in the Energy structure.

**Reply by goodkiller** [reply]
Yes, using OPTION A2.
But anyways if current is not available then it could calculate current from Power and Voltage.

**Reply by s-hadinger** [reply]
Yes but my point is that the current is set by the dummy driver according to relay position. You can't change the current youtself writing to memory because the driver will override it anyways.

Hence my point, the `energy` structure in Berry is primarily meant for reading only.

If we want a pure Berry driver for energy, we need to create a new `C` stub for it, and not use the dummy driver which has a different purpose.
Edit: similarly with what I did with Display: `xdsp_18_berry_display.ino`

**Reply by goodkiller** [reply]
Well, things are getting more confused ...
Any instructions what module to use to get energy measurements while we have only Voltage and Power? Data must be displayed on UI and in mqtt sensor section.

I thought that dummy energy driver was the right approach.

**Reply by s-hadinger** [reply]
Short answer is that it's not supported right now. It would require to develop a new energy driver in C++ or a generic one so that it can be developed in Berry. 
This should not be very complex, but not innthe following weeks

**Reply by FransOv** [reply]
It is supported, although maybe not officially. I read the values from an Eastron 360DM modbus meter and write the voltage, current and active power to the energy object. I can also write voltage current and power factor, gives about the same result but for very low current gives 0 for active power. If you only have voltage and active power, you should also set the Power Factor to 1.
You just must make sure that you do define the relays per phase, but do not turn them on. If you turn them on, the dummy energy driver will enter its dummy values.
<img width="321" alt="image" src="https://github.com/user-attachments/assets/616f3b0a-b265-4a05-a114-deb3cf0f5e35">
<img width="270" alt="image" src="https://github.com/user-attachments/assets/28783599-fd8e-4386-8d41-c408304d3a2b">
[mbmaster.be.txt](https://github.com/user-attachments/files/16512668/mbmaster.be.txt)

**Reply by sfromis** [reply]
Unrelated to the issue about setting energy monitoring data, I noticed a detail in that Berry code. Your `tasmota.remove_rule("mb_received")` will not work, as it also needs the trigger as first parameter. Using the `id` mainly benefits if there might be other places using the same trigger, where you have to specify which one to remove.

Similarly, your setup looks like you are trying to have it facilitate re-adding the code, removing the old driver. But it does not work that way, as the new instance created will be a different one from the old.
```
var pm=power_monitor()
tasmota.remove_driver(pm)
tasmota.add_driver(pm)
```
What I usually do is to have the driver setup in the `init()` function, like this:
```
tasmota.remove_driver(global.power_monitor_instance)
tasmota.add_driver(global.power_monitor_instance := self)
```
First time, the global var will be `nil` which does not cause a failure to "remove".

**Reply by FransOv** [reply]
Thanks for the tips. I was/am struggling with cleanly re-initialising. When testing I mostly resort to restarting the ESP. I will use your suggestions in the next versions.

**Reply by goodkiller** [reply]
Anyways I call vote for new generic driver as @s-hadinger proposed previously, then it is more logical in bigger picture.

**Comment by FransOv** [comment]
@s-hadinger  I'm glad that you are going to use pseudo arrays instead of the _2 _3 etc. construct. Makes for much cleaner code.

**Comment by goodkiller** [comment]
In Tasmota 14.2 energy module was refactored. Now I am not able to set current at all when "current_available" is 0. @s-hadinger what are the plans with calculating current from power and voltage?

**Reply by s-hadinger** [reply]
The Berry Energy driver is now available. See https://github.com/arendst/Tasmota/pull/22134
